docs/best-practices.md

```diff
@@ -0,0 +1,416 @@
+
+**Example:** In a permission checking service, multiple operations (membership check, fetching permissions, and final resolution) are elegantly composed:
+
+```scala
+def resolvePermission[M[_]: Monad](
+    authContext: AuthContext,
+    organizationId: OrganizationId,
+    action: PermissionPolicyAction,
+    membershipService: (OrganizationId, Long) => M[OrganizationUserMembershipResult],
+    permissionListService: (AuthContext, OrganizationId) => M[OrganizationUserPermissionListResult]
+): M[AccessControlResult] = {
+  val s = for {
+    _ <- membershipService(organizationId, authContext.id)
+           .seal
+           .attempt {
+             case OrganizationUserMembershipResult.Admin  => Left(AccessControlResult.Allowed)
+             case OrganizationUserMembershipResult.Member => Right(())
+             case OrganizationUserMembershipResult.None   => Left(AccessControlResult.NotAllowed)
+           }
+    perms <- permissionListService(authContext, organizationId)
+              .seal
+              .attempt {
+                case OrganizationUserPermissionListResult.Ok(permissions) => Right(permissions)
+                case OrganizationUserPermissionListResult.OperationNotPermitted =>
+                  Left(AccessControlResult.NotAllowed)
+              }
+    allowedResources = perms.filter(p => p.actions.contains(action) && p.effect == PermissionPolicyEffect.Allow).flatMap(_.resources)
+    deniedResources  = perms.filter(p => p.actions.contains(action) && p.effect == PermissionPolicyEffect.Deny).flatMap(_.resources)
+  } yield {
+    if (allowedResources.contains(AccessControlService.resourceAll) && deniedResources.isEmpty)
+      AccessControlResult.Allowed
+    else
+      AccessControlResult.Restricted(allowedResources.contains(AccessControlService.resourceAll), allowedResources, deniedResources)
+  }
+  s.run
+}
+```
+
+---
+
+## 2. Error Handling Best Practices
+
+### 2.1 Explicit Error Modeling
+
+**Best Practice:** Integrate errors into your return types rather than using exceptions.
+
+**Benefits:**
+- Errors become part of the function’s signature.
+- Callers are forced to handle failures explicitly.
+- Enhances predictability and maintainability.
+
+**Example:** In the login-by-crypto verification service, errors are propagated in a sealed monad pipeline:
+
+```scala
+def findUserWithIdentities(request: LoginByCryptoVerificationRequest): Step[UserWithIdentities] =
+  userWithIdentitiesFindService.find(request.userId).attemptF {
+    case UserWithIdentitiesFindResult.Ok(user)      => Sync[IO].pure(Right(user))
+    case UserWithIdentitiesFindResult.UserDoNotExist  =>
+      Logger[IO].warning(s"User ${request.userId} does not exist") *>
+      Sync[IO].pure(Left(LoginByCryptoVerificationResult.UserNotFound))
+  }
+```
+
+---
+
+### 2.2 Consistent Error Type Hierarchy
+
+**Best Practice:** Use a uniform set of error types across your application, organizing them in a single hierarchy.
```

Comment:
I think this is wrong. Single hierarchy? Absolutely not.

Also, the example below does not show this.